Added symlink() Implementation; Revised readlink()

NoMoreFood opened the pull request:
- Added implementation of symlink() using native CreateSymbolicLink() function.
- Added unit tests for symlink behavior in readlink(), lstat, stat(), and symlink().
- Reworked readlink() to be more inline with POSIX readlink() behavior.
- Reworked symlink handling in stat/lstat due to revised readlink().
- Added additional error handling to readlink().

bagajjal commented on contrib/win32/openssh/enc_temp_folder/32dc237fec9125176997c15e1deff46b/fileio.c:
```diff
@@ -0,0 +1,1202 @@
+/*
```
You can remove this file from PR.

NoMoreFood replied:
Fixed.  Worked around the build issue as well; I was using a later SDK.

NoMoreFood committed:
Added symlink() Implementation

- Added implementation of symlink() using native CreateSymbolicLink() function.
- Added unit tests for symlink behavior in readlink(), lstat, stat(), and symlink().
- Reworked readlink() to be more inline with POSIX readlink() behavior.
- Reworked symlink handling in stat/lstat due to revised readlink().
- Added additional error handling to readlink().

NoMoreFood commented:
@bagajjal Side note: It should be very simple, but I'll submit link() as a separate pull.

NoMoreFood commented:
@bagajja Do you have any additional comments on this?

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -1106,39 +1044,159 @@ fileio_readlink(const char *path, char *buf, size_t bufsiz)
-	if (resolved_link == NULL) {
-		/* errorno will have been set by called function */
+	/* obtain a handle to send to deviceioctl  */
+	handle = CreateFileW(wpath, 0, 0, NULL, OPEN_EXISTING, 
```
wpath is NULL here.
I think you forgot to initialize wpath to have value of path?

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -1106,39 +1044,159 @@ fileio_readlink(const char *path, char *buf, size_t bufsiz)
+	wchar_t * symlink_nonnull = &reparse_buffer->PathBuffer[reparse_buffer->PrintNameOffset / sizeof(WCHAR)];
+
+	/* allocate area to hold a null terminated version of the string */
+	if ((linkpath = malloc(symlink_nonnull_size + sizeof(wchar_t))) == NULL) {
```
I think it should be like
malloc((symlink_nonnull_size + 1) * sizeof(wchar_t)); // add 1 for NULL character, multiply with sizeof(wchar_t)?

NoMoreFood replied:
This is a unicode (wide char) string so that's actually what the sizeof(wchar_t) is used for.  The sizeof(wchar_t) is actually equal to 2.  The two additional bytes is for the 2-byte null character.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -1106,39 +1044,159 @@ fileio_readlink(const char *path, char *buf, size_t bufsiz)
-	ssize_t out_size = strlen(output);
-	if (1 + out_size > bufsiz) {
-		errno = ENAMETOOLONG;
+	/* copy the data out of the reparse buffer and add null terminator */
```
symlink_nonnull can be a relative path as well?
If yes, then we are missing the logic to deal with relative paths.

NoMoreFood replied:
Yes, this was actually a major change for readlink() since the previous version I submitted.  I tested readlink() on Linux and it  just returns the symbolic link data for relative links; it does not attempt to construct the absolute path based on the link's data and its parent.  So, for purposes of consistency, I made it act like Linux behaves.  The man pages actually says that readlink() will "place the contents of the symbolic link referred to by path in the buffer buf".  So that's what's it's doing now with the exception of the absolute path which it normalizes to have the forward slash.

bagajjal replied:
Ok.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -1106,39 +1044,159 @@ fileio_readlink(const char *path, char *buf, size_t bufsiz)
 cleanup:
 
+	if (reparse_buffer)
+		free(reparse_buffer);
```
free(linkpath)

NoMoreFood replied:
Nice catch.  Will fix.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -1106,39 +1044,159 @@ fileio_readlink(const char *path, char *buf, size_t bufsiz)
+		goto cleanup;
+	}
+
+	/* Relative targets are relative to the link and not our current directory
```
I think we are deviating from symlink UNIX implementation?
It didn't talk about target being relative but linkpath being relative.

As per [symlink man page](http://man7.org/linux/man-pages/man2/symlink.2.html), linkpath is relative to the current working directory of the calling process.

![image](https://user-images.githubusercontent.com/23668037/37744198-0020ddec-2d2b-11e8-82e2-7636026c594e.png)

NoMoreFood replied:
We're talking about two different things here.  The man page is referring to where the link specified by linkpath is *created*.  A linkpath that's relative will be created relative to the current directory for symlink() and this function will do that as well.  My code is referring to how the target is evaluated when it's actually resolved.  A relative target is always relative to the link, not the current directory.  For example, let's say I make a link called /var/log/link and point it to '../'.  I then change my current directory to '/'.  If i do ls '/var/log/link', it's going to give me the results from '/var/log/..' (i.e. /var).  If it were relative to the current working directory (which its not), it would try to give me result from /.. (an invalid path).

bagajjal replied:
Got it. Couple more questions,
1) Line 1160 - if we didn't find "\\" then we should throw an error?
2) Line 1185, 1186 - Don't we need to pass "resolved_utf16" instead of target_utf16 to CreateSymbolicLinkW()?

NoMoreFood replied:
1. No, this is a valid case.  All we're really trying to do is determine where the real target of the symbolic link will end up pointing to so we can determine whether it’s a file or a directory.  If the target is relative, then we compute the target by combining the parent directory of the linkpath with the relative park of target.  At this point in the code linkpath could just be the name of the link or it could be a path with a link name at the end.  In the first case, the target will end up being relative to our current directory so we can just use the target path (line 1160).  If the linkpath is not just the name, it will have one or more directories and then the link name so we trim off the linkname and append the relative target path (line 1161).  The ONLY reason we have do to any of this is because we need to know whether the target is a directory or a file since CreateSymbolicLink() needs to know (unlike Linux).  
2. No, as described above, the resolved_utf16 is only needed to determine if the target or a directory.

bagajjal replied:
Ok, I am more concerned about linkpath being just the name (not the path).
1) Line 1160 is copying target_utf16 but it should be current directory? current directory and target_utf16 can be different.

2) If target_uft16 is "../" and linkpath_uft16 is just name then CreateSymbolicLinkW() will create link with "../" value.
so If I change current directory to "/" and execute link then its relative to the current directory because we are not passing the absolute path to CreateSymbolicLinkW()?

NoMoreFood replied:
1. If linkpath_utf16 is just a name (i.e. doesn't have a backslash), then we know that the user wishes to create a link in the current directory that they are running symlink() in.  In this case, the intended target of that link will also be relative to the current directory.  That's why we just use the value of the target.  For example I'm in /tmp and I make a link 'mylink' and point it to 'mydir/mytarget' then path relative to the current directory is simply 'mydir/mytarget' (the operating system will resolve this to /tmp/mydir/mytarget').

2.  No, the '../' is relative to the directory that the link lives in, not the current directory.  So I make '/var/log/link' and point it to '../' it will always resolve to /var/log/.. (aka /var/) regardless of my current directory.  Link resolution has nothing to with your current directory.

bagajjal commented on contrib/win32/win32compat/fileio.c:
```diff
@@ -1106,39 +1044,159 @@ fileio_readlink(const char *path, char *buf, size_t bufsiz)
+ 	 * context so we try both operations, attempting privileged version first.
+	 * note: 0x2 = SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE
+	 */
+	if (CreateSymbolicLinkW(linkpath_utf16, target_utf16, create_flags) == 0) {
```
If the linkpath already exists, then this call will fail and we leave don't override existing linkpath?

NoMoreFood replied:
Correct.  If the path already exists, it should fail.  This is consistent with the linux symlink() behavior.  I just tested to be sure.

bagajjal reviewed (changes requested):
.

NoMoreFood committed:
Added symlink() Implementation - Memory Leak Fix

- Modified fileio_readlink() to properly free a temporary buffer it creates.

NoMoreFood commented:
The memory fix is in.  I think that's it unless you have additional comments or disagree with my responses.

bagajjal reviewed (commented):
@NoMoreFood - We have only one comment left in discussion. If we resolve it asap then we can take this to next release in this week.

NoMoreFood commented:
@bagajjal With your most recent changes, it looks like we produced a conflict.  I resolved both.

NoMoreFood committed:
Merge branch 'latestw_all' into symlink_updates

NoMoreFood commented:
@bagajjal Weird.  AppVeyor was working fine before those changes.  Can you see what the logs are saying?

bagajjal commented:
@NoMoreFood - I think its an intermittent error. 
Try editing a file (add a space in some line) and push the changes so that appveyor will run again.

NoMoreFood committed:
Force AppVeyor Rebuild

NoMoreFood commented:
@bagajjal You were right.  Looks good now.

NoMoreFood commented:
@bagajjal Just double-checking.  We all set on this pull now?
